File: backend/utils/recaptcha.py
"""
Helper to validate Google reCAPTCHA tokens.
"""

from typing import Optional
import os

import requests

RECAPTCHA_SECRET_KEY = os.getenv("RECAPTCHA_SECRET_KEY")
RECAPTCHA_VERIFY_URL = "https://www.google.com/recaptcha/api/siteverify"


def verify_recaptcha_token(token: Optional[str]) -> bool:
    """Verify the token against Google only when a secret key exists."""
    if not RECAPTCHA_SECRET_KEY:
        return True
    if not token:
        return False

    try:
        response = requests.post(
            RECAPTCHA_VERIFY_URL,
            data={"secret": RECAPTCHA_SECRET_KEY, "response": token},
            timeout=5,
        )
        response.raise_for_status()
        payload = response.json()
        return bool(payload.get("success"))
    except requests.RequestException:
        return False

File: frontend/src/lib/recaptcha.ts
"use server";

export async function verifyRecaptchaToken(token: unknown): Promise<{
  ok: boolean;
  status?: number;
  error?: string;
}> {
  const recaptchaSecret = process.env.RECAPTCHA_SECRET_KEY;
  const isDev = process.env.NODE_ENV === "development";

  if (!recaptchaSecret || isDev) {
    // reCAPTCHA not configured or dev environment — skip verification
    return { ok: true };
  }

  if (typeof token !== "string" || !token.trim()) {
    // Client couldn't load reCAPTCHA script (e.g. domain not registered, ad blocker) — allow through
    return { ok: true };
  }

  try {
    const recaptchaResponse = await fetch("https://www.google.com/recaptcha/api/siteverify", {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: new URLSearchParams({
        secret: recaptchaSecret,
        response: token,
      }),
    });
    const recaptchaResult = (await recaptchaResponse.json().catch(() => null)) as
      | { success?: boolean; score?: number }
      | null;
    if (!recaptchaResult?.success || (typeof recaptchaResult.score === "number" && recaptchaResult.score < 0.5)) {
      return { ok: false, status: 400, error: "reCAPTCHA failed" };
    }
  } catch {
    return { ok: false, status: 500, error: "reCAPTCHA verification service unavailable" };
  }

  return { ok: true };
}
